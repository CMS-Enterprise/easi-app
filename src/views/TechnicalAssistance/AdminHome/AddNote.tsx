import React, { useContext, useEffect } from 'react';
import { Controller, useForm } from 'react-hook-form';
import { Trans, useTranslation } from 'react-i18next';
import { useHistory, useParams } from 'react-router-dom';
import { useMutation } from '@apollo/client';
import {
  Button,
  Dropdown,
  ErrorMessage,
  Form,
  FormGroup,
  Grid,
  IconArrowBack,
  Label
} from '@trussworks/react-uswds';
import classNames from 'classnames';

import UswdsReactLink from 'components/LinkWrapper';
import PageHeading from 'components/PageHeading';
import Alert from 'components/shared/Alert';
import { ErrorAlertMessage } from 'components/shared/ErrorAlert';
import TextAreaField from 'components/shared/TextAreaField';
import useMessage from 'hooks/useMessage';
import CreateTrbAdminNote from 'queries/CreateTrbAdminNote';
import { TRBAdminNoteFragment } from 'queries/GetTrbAdminNotesQuery';
import {
  CreateTrbAdminNote as CreateTrbAdminNoteType,
  CreateTrbAdminNoteVariables
} from 'queries/types/CreateTrbAdminNote';
import {
  CreateTRBAdminNoteInput,
  TRBAdminNoteCategory
} from 'types/graphql-global-types';

import Breadcrumbs from '../Breadcrumbs';

import { ModalViewType } from './components/NoteModal';
import { TRBRequestContext } from './RequestContext';

const AddNote = ({
  trbRequestId,
  setModalView, // prop used to conditionall render text/links/etc specifically for modal
  setModalMessage,
  defaultSelect
}: {
  trbRequestId?: string;
  setModalView?: React.Dispatch<React.SetStateAction<ModalViewType>>;
  setModalMessage?: React.Dispatch<React.SetStateAction<string>>;
  defaultSelect?: TRBAdminNoteCategory;
}) => {
  const { t } = useTranslation('technicalAssistance');

  const { id } = useParams<{
    id: string;
  }>();

  const history = useHistory();

  // TRB request information to render name in breadcrumbs
  const { data } = useContext(TRBRequestContext);

  const { message, showMessage, showMessageOnNextPage } = useMessage();

  const requestUrl: string = `/trb/${id}/notes`;

  const [mutate] = useMutation<
    CreateTrbAdminNoteType,
    CreateTrbAdminNoteVariables
  >(CreateTrbAdminNote, {
    update(cache, { data: modifiedData }) {
      cache.modify({
        id: cache.identify({ __typename: 'TRBRequest', id }),
        fields: {
          adminNotes(existingNotes = []) {
            const newNote = cache.writeFragment({
              data: modifiedData?.createTRBAdminNote,
              fragment: TRBAdminNoteFragment
            });
            return [...existingNotes, newNote];
          }
        }
      });
    }
  });

  const defaultValues: Omit<CreateTRBAdminNoteInput, 'trbRequestId'> = {
    category: defaultSelect || ('' as TRBAdminNoteCategory),
    noteText: ''
  };

  const {
    control,
    handleSubmit,
    formState: { errors, dirtyFields, isSubmitting }
  } = useForm({
    defaultValues
  });

  const hasErrors: boolean = Object.keys(errors).length > 0;

  useEffect(() => {
    if (hasErrors) {
      const err = document.querySelector('.trb-basic-fields-error');
      err?.scrollIntoView();
    }
  }, [errors, hasErrors]);

  return (
    <div
      className={classNames({
        'grid-container': !setModalView
      })}
    >
      {!setModalView && (
        <Breadcrumbs
          items={[
            { text: t('Home'), url: `/trb` },
            {
              text: t('adminHome.breadcrumb', {
                trbRequestId: data?.trbRequest.name || ''
              }),
              url: requestUrl
            },
            {
              text: t('notes.addNote')
            }
          ]}
        />
      )}

      <Form
        onSubmit={handleSubmit(formData => {
          mutate({
            variables: {
              input: {
                trbRequestId: trbRequestId || id, // Get id from prop(modal) or url param
                category: formData.category,
                noteText: formData.noteText
              }
            }
          })
            .then(result => {
              if (!setModalView) {
                showMessageOnNextPage(
                  <Alert type="success" className="margin-top-3">
                    {t('notes.status.success')}
                  </Alert>
                );
                history.push(requestUrl);
              } else if (setModalView && setModalMessage) {
                setModalView('viewNotes');
                setModalMessage(t('notes.status.success'));
              }
            })
            .catch(err => {
              showMessage(
                <Alert type="error" className="margin-top-3">
                  {t('notes.status.error')}
                </Alert>
              );
            });
        })}
        className={classNames('maxw-full', {
          'desktop:grid-col-6': !setModalView
        })}
      >
        <Grid row>
          <Grid col>
            <PageHeading
              className={classNames('margin-bottom-0', {
                'margin-top-0': setModalView
              })}
            >
              {t('notes.addNote')}
            </PageHeading>
            <div className="line-height-body-5 font-body-lg text-light">
              {t('notes.addNoteDescription')}
            </div>
            <div className="margin-top-1 margin-bottom-2 text-base">
              <Trans
                i18nKey="technicalAssistance:actionRequestEdits.fieldsMarkedRequired"
                components={{ red: <span className="text-red" /> }}
              />
            </div>

            {message}

            {hasErrors && (
              <Alert
                heading={t('errors.checkFix')}
                type="error"
                slim={false}
                className="trb-basic-fields-error"
              >
                {Object.keys(errors).map(fieldName => {
                  const msg: string = t(`notes.labels.${fieldName}`);
                  return (
                    <ErrorAlertMessage
                      key={fieldName}
                      errorKey={fieldName}
                      message={msg}
                    />
                  );
                })}
              </Alert>
            )}
          </Grid>
        </Grid>

        <Grid row gap>
          <Grid desktop={{ col: 12 }}>
            {/* Category */}

            <Controller
              name="category"
              control={control}
              rules={{ required: true }}
              render={({ field, fieldState: { error } }) => (
                <FormGroup error={!!error}>
                  <Label
                    htmlFor="category"
                    className="text-normal"
                    error={!!error}
                  >
                    {t('notes.labels.category')}{' '}
                    <span className="text-red">*</span>
                  </Label>

                  {error && (
                    <ErrorMessage>{t('errors.fillBlank')}</ErrorMessage>
                  )}

                  <Dropdown
                    id="category"
                    data-testid="note-category"
                    {...field}
                    ref={null}
                  >
                    <option key="default-select" disabled value="">
                      - {t('basic.options.select')} -
                    </option>

                    {[
                      TRBAdminNoteCategory.GENERAL_REQUEST,
                      TRBAdminNoteCategory.INITIAL_REQUEST_FORM,
                      TRBAdminNoteCategory.SUPPORTING_DOCUMENTS,
                      TRBAdminNoteCategory.CONSULT_SESSION,
                      TRBAdminNoteCategory.ADVICE_LETTER
                    ].map(key => (
                      <option key={key} value={key}>
                        {t(`notes.categories.${key}`)}
                      </option>
                    ))}
                  </Dropdown>
                </FormGroup>
              )}
            />

            {/* Note Text */}
            <Controller
              name="noteText"
              control={control}
              render={({ field, fieldState: { error } }) => (
                <FormGroup>
                  <Label htmlFor="noteText" className="text-normal">
                    {t('notes.labels.noteText')}
                    <span className="text-red">*</span>
                  </Label>

                  <TextAreaField
                    {...field}
                    ref={null}
                    id={field.name}
                    error={!!error}
                  />
                </FormGroup>
              )}
            />
          </Grid>
        </Grid>

        <div className="margin-top-3">
          {setModalView && (
            <Button
              type="button"
              className="usa-button--outline"
              onClick={() => {
                setModalView('viewNotes');
              }}
            >
              {t('notes.cancel')}
            </Button>
          )}
          <Button
            type="submit"
            name={t('notes.saveNote')}
            disabled={
              Object.keys(dirtyFields).length !==
                Object.keys(defaultValues).length || isSubmitting
            }
          >
            {t('notes.saveNote')}
          </Button>
        </div>
      </Form>

      {!setModalView && (
        <div className="margin-top-2">
          <UswdsReactLink to={requestUrl}>
            <IconArrowBack className="margin-right-05 margin-bottom-2px text-tbottom" />
            {t('actionRequestEdits.cancelAndReturn')}
          </UswdsReactLink>
        </div>
      )}
    </div>
  );
};

export default AddNote;
