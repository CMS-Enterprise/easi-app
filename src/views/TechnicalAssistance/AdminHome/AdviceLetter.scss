.trb-admin-home__advice .easi-pdf-export__controls {
  margin-top: 0 !important;
}
