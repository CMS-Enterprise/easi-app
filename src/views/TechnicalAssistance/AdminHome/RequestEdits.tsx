import React from 'react';
import { Controller } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import { useHistory, useParams } from 'react-router-dom';
import { useMutation } from '@apollo/client';
import { yupResolver } from '@hookform/resolvers/yup';
import { ErrorMessage, FormGroup } from '@trussworks/react-uswds';

import HelpText from 'components/shared/HelpText';
import Label from 'components/shared/Label';
import TextAreaField from 'components/shared/TextAreaField';
import useMessage from 'hooks/useMessage';
import CreateTrbRequestFeedbackQuery from 'queries/CreateTrbRequestFeedbackQuery';
import {
  CreateTrbRequestFeedback,
  CreateTrbRequestFeedbackVariables
} from 'queries/types/CreateTrbRequestFeedback';
import { TRBFeedbackAction } from 'types/graphql-global-types';
import { TrbRecipientFields } from 'types/technicalAssistance';
import { trbActionSchema } from 'validations/trbRequestSchema';

// import Breadcrumbs from '../Breadcrumbs';
import useActionForm from './components/ActionFormWrapper/useActionForm';

interface RequestEditsFields extends TrbRecipientFields {
  feedbackMessage: string;
}

function RequestEdits() {
  const { t } = useTranslation('technicalAssistance');

  const { id, activePage, action } = useParams<{
    id: string;
    activePage: string;
    action: 'request-edits' | 'ready-for-consult';
  }>();
  const history = useHistory();

  const { showMessage, showMessageOnNextPage } = useMessage();

  const requestUrl = `/trb/${id}/${activePage}`;

  let actionText: 'actionRequestEdits' | 'actionReadyForConsult';
  let feedbackAction: TRBFeedbackAction;

  if (action === 'request-edits') {
    actionText = 'actionRequestEdits';
    feedbackAction = TRBFeedbackAction.REQUEST_EDITS;
  } else {
    actionText = 'actionReadyForConsult';
    feedbackAction = TRBFeedbackAction.READY_FOR_CONSULT;
  }

  const {
    control,
    ActionForm,
    handleSubmit,
    watch,
    formState: { isSubmitting }
  } = useActionForm<RequestEditsFields>({
    trbRequestId: id,
    resolver: yupResolver(
      trbActionSchema(
        'feedbackMessage',
        feedbackAction === TRBFeedbackAction.REQUEST_EDITS
      )
    ),
    defaultValues: {
      feedbackMessage: '',
      copyTrbMailbox: true,
      notifyEuaIds: []
    }
  });

  const [sendFeedback, feedbackResult] = useMutation<
    CreateTrbRequestFeedback,
    CreateTrbRequestFeedbackVariables
  >(CreateTrbRequestFeedbackQuery);

  const submitForm = (formData: RequestEditsFields) => {
    sendFeedback({
      variables: {
        input: { ...formData, trbRequestId: id, action: feedbackAction }
      }
    })
      .then(result => {
        showMessageOnNextPage(t(`${actionText}.success`), {
          type: 'success',
          className: 'margin-top-3'
        });
        history.push(requestUrl);
      })
      .catch(err => {
        showMessage(t(`${actionText}.error`), {
          type: 'error',
          className: 'margin-top-3'
        });
      });
  };

  return (
    <ActionForm
      title={t(`${actionText}.heading`)}
      description={t(`${actionText}.description`)}
      onSubmit={handleSubmit(formData => submitForm(formData))}
      buttonProps={{
        next: {
          text: t('actionRequestEdits.submit'),
          disabled:
            feedbackAction === TRBFeedbackAction.REQUEST_EDITS &&
            !watch('feedbackMessage'),
          loading: isSubmitting || feedbackResult.loading
        },
        taskListUrl: requestUrl,
        saveExitText: t('actionRequestEdits.cancelAndReturn')
      }}
      breadcrumbItems={[
        {
          text: t(`${actionText}.breadcrumb`)
        }
      ]}
    >
      <Controller
        name="feedbackMessage"
        control={control}
        render={({ field, fieldState: { error } }) => {
          return (
            <FormGroup error={!!error}>
              <Label
                htmlFor="feedbackMessage"
                className="text-normal margin-top-6"
                required={feedbackAction === TRBFeedbackAction.REQUEST_EDITS}
              >
                {t(`${actionText}.label`)}
              </Label>
              <HelpText id="feedbackmessage-hint">
                {t('actionRequestEdits.hint')}
              </HelpText>
              {error && <ErrorMessage>{t('errors.fillBlank')}</ErrorMessage>}
              <TextAreaField
                id="feedbackMessage"
                {...field}
                ref={null}
                aria-describedby="feedbackMessage-info feedbackMessage-hint"
              />
            </FormGroup>
          );
        }}
      />
    </ActionForm>
  );
}

export default RequestEdits;
