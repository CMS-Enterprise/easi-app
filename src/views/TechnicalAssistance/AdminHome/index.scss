@use 'uswds-core' as *;

.trb-admin__view-all-link a {
  display: flex;
  align-items: center;
  color: primary;

  &:hover {
    color: color("primary-dark");
  }

  svg {
    margin-right: .5rem;
  }
}

.trb-admin__nav-link {
  &--border {
    border-bottom: 1px solid color('disabled-light');
  }

  a {
    display: block;
    text-decoration: none;
    padding: .5rem 0;
    color: color('base-darkest');

    &:hover {
      color: color('primary');
      background-color: color('base-lightest');
    }

    span {
      padding: .25rem 0;
    }
  }

  &--active a {
    color: color('primary');

    span {
      border-left: 3px solid color('primary');
      padding-left: 1rem;
      font-weight: bold;
    }
  }
}
