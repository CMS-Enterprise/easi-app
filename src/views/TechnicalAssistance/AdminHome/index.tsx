import React, { useContext, useMemo, useRef } from 'react';
import { useTranslation } from 'react-i18next';
import { useSelector } from 'react-redux';
import { Link, Route, useParams } from 'react-router-dom';
import {
  Grid,
  GridContainer,
  IconArrowBack,
  ModalRef
} from '@trussworks/react-uswds';
import classNames from 'classnames';
import { useFlags } from 'launchdarkly-react-client-sdk';

import PageLoading from 'components/PageLoading';
import cmsDivisionsAndOffices from 'constants/enums/cmsDivisionsAndOffices';
import useMessage from 'hooks/useMessage';
import useTRBAttendees from 'hooks/useTRBAttendees';
import { AppState } from 'reducers/rootReducer';
import { TrbRequestIdRef } from 'types/technicalAssistance';
import { formatDateLocal } from 'utils/date';
import user from 'utils/user';
import AccordionNavigation from 'views/GovernanceReviewTeam/AccordionNavigation';
import NotFound from 'views/NotFound';

import Summary from './components/Summary';
import { TRBRequestContext } from './RequestContext';
import trbAdminPages from './trbAdminPages';
import TrbAssignLeadModal from './TrbAssignLeadModal';

import './index.scss';

type SideNavProps = {
  activePage: string;
  trbRequestId: string;
};

/** Side navigation */
const SideNavigation = ({
  /** Active page path */
  activePage,
  /** Request ID */
  trbRequestId
}: SideNavProps) => {
  const { t } = useTranslation('technicalAssistance');
  return (
    <nav>
      <ul className="trb-admin__nav-list usa-list usa-list--unstyled">
        <li className="trb-admin__view-all-link margin-bottom-4">
          <Link to="/">
            <IconArrowBack aria-hidden />
            {t('adminHome.backToRequests')}
          </Link>
        </li>
        {trbAdminPages.map(({ path, text, groupEnd }) => {
          return (
            <li
              key={text}
              className={classNames('trb-admin__nav-link', {
                'trb-admin__nav-link--active': path === activePage,
                'trb-admin__nav-link--border': groupEnd
              })}
            >
              <Link to={`/trb/${trbRequestId}/${path}`}>
                <span>{t(text)}</span>
              </Link>
            </li>
          );
        })}
      </ul>
    </nav>
  );
};

/** Wrapper for TRB admin view components */
export default function AdminHome() {
  // Current user info from redux
  const { groups, isUserSet } = useSelector((state: AppState) => state.auth);

  const { t } = useTranslation('technicalAssistance');

  const flags = useFlags();

  // Get url params
  const { id, activePage } = useParams<{
    id: string;
    activePage: string;
  }>();

  const trbContextData = useContext(TRBRequestContext);

  const { data, loading } = trbContextData;

  const trbRequest = data?.trbRequest;

  // Alert feedback from children
  const { message } = useMessage();

  // Get requester object from request attendees
  const {
    data: { requester, loading: requesterLoading }
  } = useTRBAttendees(id);

  /**
   * Requester name and cms office acronym
   */
  const requesterString = useMemo(() => {
    // If loading, return null
    if (requesterLoading) return null;

    // If requester component is not set, return name or EUA
    if (!requester.component)
      return requester?.userInfo?.commonName || requester?.userInfo?.euaUserId;

    /** Requester component */
    const requesterComponent = cmsDivisionsAndOffices.find(
      object => object.name === requester.component
    );

    // Return requester name and component acronym
    return `${requester?.userInfo?.commonName}, ${requesterComponent?.acronym}`;
  }, [requester, requesterLoading]);

  // Note count for NoteBox modal rendered on each page
  // const noteCount: number = (data?.trbRequest?.adminNotes || []).length;

  // Assign trb lead modal refs
  const assignLeadModalRef = useRef<ModalRef>(null);
  const assignLeadModalTrbRequestIdRef = useRef<TrbRequestIdRef>(null);

  // If TRB request is loading or user is not set, return page loading
  if (loading || !isUserSet) {
    return <PageLoading />;
  }

  // If TRB request does not exist or user is not TRB admin, return page not found
  if (!trbRequest || !user.isTrbAdmin(groups, flags)) {
    return <NotFound />;
  }

  const submissionDate = formatDateLocal(trbRequest.createdAt, 'MMMM d, yyyy');

  return (
    <div id="trbAdminHome">
      {/* Request summary */}
      <Summary
        trbRequestId={id}
        name={trbRequest.name || t('taskList.defaultName')}
        requestType={trbRequest.type}
        state={trbRequest.state}
        taskStatus={trbRequest.status}
        trbLead={trbRequest.trbLeadInfo.commonName}
        requester={requester}
        requesterString={requesterString}
        submissionDate={submissionDate}
        assignLeadModalRef={assignLeadModalRef}
        assignLeadModalTrbRequestIdRef={assignLeadModalTrbRequestIdRef}
      />

      {/* Accordion navigation for tablet and mobile */}
      <AccordionNavigation
        activePage={activePage}
        subNavItems={trbAdminPages.map(({ path, text, groupEnd }) => ({
          text,
          route: `/trb/${id}/${path}`,
          groupEnd
        }))}
        defaultTitle="TRB Request"
      />

      <GridContainer>
        {message}
        <Grid row className="margin-top-5 grid-gap">
          {/* Side navigation */}
          <Grid
            col
            desktop={{ col: 3 }}
            className="display-none desktop:display-block"
          >
            <SideNavigation activePage={activePage} trbRequestId={id} />
          </Grid>

          {/* Page component */}
          <Grid col desktop={{ col: 9 }}>
            {trbAdminPages.map(subpage => (
              <Route
                exact
                path={`/trb/${id}/${subpage.path as string}`}
                key={subpage.path}
              >
                <subpage.component
                  trbRequestId={id}
                  trbRequest={trbRequest}
                  requesterString={requesterString}
                  assignLeadModalRef={assignLeadModalRef}
                  assignLeadModalTrbRequestIdRef={
                    assignLeadModalTrbRequestIdRef
                  }
                />
              </Route>
            ))}
          </Grid>
        </Grid>
      </GridContainer>

      <TrbAssignLeadModal
        modalRef={assignLeadModalRef}
        trbRequestIdRef={assignLeadModalTrbRequestIdRef}
      />
    </div>
  );
}
