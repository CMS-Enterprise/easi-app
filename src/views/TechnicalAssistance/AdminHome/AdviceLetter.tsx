import React from 'react';
import { useTranslation } from 'react-i18next';

import PageLoading from 'components/PageLoading';
import Alert from 'components/shared/Alert';
import useCacheQuery from 'hooks/useCacheQuery';
import { GetTrbAdviceLetterQuery } from 'queries/TrbAdviceLetterQueries';
import {
  GetTrbAdviceLetter,
  GetTrbAdviceLetterVariables
} from 'queries/types/GetTrbAdviceLetter';
import { TRBAdviceLetterStatus } from 'types/graphql-global-types';
import { TrbAdminPageProps } from 'types/technicalAssistance';

import ReviewAdviceLetter from './components/ReviewAdviceLetter';
import TrbAdminWrapper from './components/TrbAdminWrapper';

import './AdviceLetter.scss';

const AdviceLetter = ({
  trbRequest,
  assignLeadModalTrbRequestIdRef,
  assignLeadModalRef
}: TrbAdminPageProps) => {
  const { t } = useTranslation('technicalAssistance');

  const { id } = trbRequest;

  // TRB request query
  const { data, loading } = useCacheQuery<
    GetTrbAdviceLetter,
    GetTrbAdviceLetterVariables
  >(GetTrbAdviceLetterQuery, {
    variables: { id }
  });

  const { adviceLetter, taskStatuses } = data?.trbRequest || {};

  const adviceLetterStatus = taskStatuses?.adviceLetterStatus;

  const author = adviceLetter?.author;

  // Page loading
  if (loading) return <PageLoading />;

  return (
    <TrbAdminWrapper
      activePage="advice"
      trbRequestId={id}
      title={t('adminHome.adviceLetter')}
      description={t('adviceLetter.introText')}
      disableStep={
        taskStatuses?.adviceLetterStatus ===
        TRBAdviceLetterStatus.CANNOT_START_YET
      }
      statusTagProps={{
        status:
          taskStatuses?.adviceLetterStatus ||
          TRBAdviceLetterStatus.CANNOT_START_YET,
        name: author?.commonName!,
        date: adviceLetter?.modifiedAt || adviceLetter?.createdAt || ''
      }}
      noteCount={trbRequest.adminNotes.length}
      adminActionProps={{
        status: trbRequest.status,
        state: trbRequest.state,
        assignLeadModalTrbRequestIdRef,
        assignLeadModalRef
      }}
    >
      {
        // If advice letter status is CANNOT_START_YET, show alert message
        adviceLetterStatus === 'CANNOT_START_YET' ? (
          <Alert type="info" slim>
            {t('adviceLetter.alerts.info')}
          </Alert>
        ) : (
          <>
            {adviceLetter && (
              <ReviewAdviceLetter
                adviceLetter={adviceLetter}
                trbRequestId={trbRequest.id}
                publicForm
              />
            )}
          </>
        )
      }
    </TrbAdminWrapper>
  );
};

export default AdviceLetter;
