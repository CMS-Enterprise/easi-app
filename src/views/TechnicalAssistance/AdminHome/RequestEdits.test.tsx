import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter, Route } from 'react-router-dom';
import { MockedProvider } from '@apollo/client/testing';
import { render, waitForElementToBeRemoved } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import i18next from 'i18next';

import {
  getTrbLeadOptionsQuery,
  getTRBRequestAttendeesQuery,
  getTrbRequestQuery,
  getTrbRequestSummaryQuery,
  requester,
  trbRequestSummary
} from 'data/mock/trbRequest';
import { MessageProvider } from 'hooks/useMessage';
import CreateTrbRequestFeedbackQuery from 'queries/CreateTrbRequestFeedbackQuery';
import {
  CreateTrbRequestFeedback,
  CreateTrbRequestFeedbackVariables
} from 'queries/types/CreateTrbRequestFeedback';
import { TRBFeedbackAction } from 'types/graphql-global-types';
import { MockedQuery } from 'types/util';
import easiMockStore from 'utils/testing/easiMockStore';
import VerboseMockedProvider from 'utils/testing/VerboseMockedProvider';

import TRBRequestInfoWrapper from './RequestContext';
import RequestEdits from './RequestEdits';
import AdminHome from '.';

describe('Trb Admin: Action: Request Edits', () => {
  const store = easiMockStore({ groups: ['EASI_TRB_ADMIN_D'] });

  const trbRequestId = trbRequestSummary.id;
  const feedbackMessage = 'test message';

  const createTrbRequestFeedbackQuery: MockedQuery<
    CreateTrbRequestFeedback,
    CreateTrbRequestFeedbackVariables
  > = {
    request: {
      query: CreateTrbRequestFeedbackQuery,
      variables: {
        input: {
          trbRequestId,
          feedbackMessage,
          copyTrbMailbox: true,
          notifyEuaIds: [requester.userInfo.euaUserId],
          action: TRBFeedbackAction.REQUEST_EDITS
        }
      }
    },
    result: {
      data: {
        createTRBRequestFeedback: {
          id: '94ebed72-8c66-41fd-aaa6-085a715737c2',
          __typename: 'TRBRequestFeedback'
        }
      }
    }
  };

  it('submits a feedback message', async () => {
    const {
      getByLabelText,
      getByRole,
      findByText,
      asFragment,
      getByTestId
    } = render(
      <Provider store={store}>
        <VerboseMockedProvider
          defaultOptions={{
            watchQuery: { fetchPolicy: 'no-cache' },
            query: { fetchPolicy: 'no-cache' }
          }}
          mocks={[
            createTrbRequestFeedbackQuery,
            getTrbRequestQuery,
            getTrbLeadOptionsQuery,
            getTRBRequestAttendeesQuery,
            getTRBRequestAttendeesQuery,
            getTrbRequestSummaryQuery,
            getTrbRequestSummaryQuery
          ]}
        >
          <MemoryRouter
            initialEntries={[
              `/trb/${trbRequestId}/initial-request-form/request-edits`
            ]}
          >
            <TRBRequestInfoWrapper>
              <MessageProvider>
                <Route exact path="/trb/:id/:activePage">
                  <AdminHome />
                </Route>
                <Route exact path="/trb/:id/:activePage/:action">
                  <RequestEdits />
                </Route>
              </MessageProvider>
            </TRBRequestInfoWrapper>
          </MemoryRouter>
        </VerboseMockedProvider>
      </Provider>
    );

    await waitForElementToBeRemoved(() => getByTestId('page-loading'));

    await findByText(
      i18next.t<string>('technicalAssistance:actionRequestEdits.heading')
    );

    expect(asFragment()).toMatchSnapshot();

    const submitButton = getByRole('button', {
      name: i18next.t<string>('technicalAssistance:actionRequestEdits.submit')
    });

    userEvent.type(
      getByLabelText(
        RegExp(
          i18next.t<string>('technicalAssistance:actionRequestEdits.label')
        )
      ),
      feedbackMessage
    );

    userEvent.click(submitButton);

    await findByText(
      i18next.t<string>('technicalAssistance:actionRequestEdits.success')
    );
  });

  it('shows error notice when submission fails', async () => {
    const { getByLabelText, getByRole, findByText } = render(
      <MockedProvider
        mocks={[
          getTRBRequestAttendeesQuery,
          {
            ...createTrbRequestFeedbackQuery,
            error: new Error()
          }
        ]}
      >
        <MemoryRouter
          initialEntries={[
            `/trb/${trbRequestId}/initial-request-form/request-edits`
          ]}
        >
          <TRBRequestInfoWrapper>
            <MessageProvider>
              <Route exact path="/trb/:id/:activePage/:action">
                <RequestEdits />
              </Route>
            </MessageProvider>
          </TRBRequestInfoWrapper>
        </MemoryRouter>
      </MockedProvider>
    );

    await findByText(
      i18next.t<string>('technicalAssistance:actionRequestEdits.heading')
    );

    userEvent.type(
      getByLabelText(
        RegExp(
          i18next.t<string>('technicalAssistance:actionRequestEdits.label')
        )
      ),
      feedbackMessage
    );

    userEvent.click(
      getByRole('button', {
        name: i18next.t<string>('technicalAssistance:actionRequestEdits.submit')
      })
    );

    await findByText(
      i18next.t<string>('technicalAssistance:actionRequestEdits.error')
    );
  });
});
