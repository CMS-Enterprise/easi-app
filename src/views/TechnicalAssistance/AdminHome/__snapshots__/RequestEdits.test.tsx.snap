// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trb Admin: Action: Request Edits submits a feedback message 1`] = `
<DocumentFragment>
  <div
    class="grid-container width-full"
    data-testid="gridContainer"
  >
    <nav
      aria-label="Breadcrumbs"
      class="usa-breadcrumb padding-bottom-0"
      style="background-color: transparent;"
    >
      <ol
        class="usa-breadcrumb__list"
      >
        <li
          class="usa-breadcrumb__list-item"
        >
          <a
            class="usa-breadcrumb__link"
            href="/trb"
          >
            <span>
              Home
            </span>
          </a>
        </li>
        <li
          class="usa-breadcrumb__list-item"
        >
          <a
            class="usa-breadcrumb__link"
            href="/trb/d8ebedca-0031-4ccd-9690-37390726c50e/initial-request-form"
          >
            <span>
              Request details
            </span>
          </a>
        </li>
        <li
          aria-current="page"
          class="usa-breadcrumb__list-item usa-current"
        >
          <span>
            Request edits
          </span>
        </li>
      </ol>
    </nav>
    <h1
      aria-live="polite"
      class="easi-h1 margin-bottom-0"
      tabindex="-1"
    >
      Action: request edits
    </h1>
    <p
      class="line-height-body-5 font-body-lg text-light margin-0"
    >
      Use this action if the TRB needs additional information about the request in order to proceed with scheduling a consult session. Specify the edits or additional information needed from the requester.
    </p>
    <p
      class="margin-top-1 text-base"
    >
      Fields marked with an asterisk ( 
      <span
        class="text-red"
      >
        *
      </span>
       ) are required.
    </p>
    <form
      class="usa-form maxw-none tablet:grid-col-12 desktop:grid-col-6"
      data-testid="form"
    >
      <div
        class="usa-form-group"
        data-testid="formGroup"
      >
        <label
          class="usa-label text-normal margin-top-6"
          for="feedbackMessage"
        >
          What type of edits are needed prior to scheduling the consult session?
          <span
            class="text-red"
          >
             *
          </span>
        </label>
        <div
          class="easi-help-text"
          id="feedbackmessage-hint"
        >
          Provide feedback to the requester about the content of their initial request form or supporting documents. The requester will see this feedback in their task list and in the email you send.
        </div>
        <textarea
          aria-describedby="feedbackMessage-info feedbackMessage-hint"
          class="usa-textarea"
          id="feedbackMessage"
          name="feedbackMessage"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          class="easi-character-counter margin-top-1"
          id="feedbackMessage-counter"
        >
          2000 characters remaining
        </span>
      </div>
      <h3
        class="margin-top-6 margin-bottom-0"
      >
        Notification email
      </h3>
      <p
        class="margin-0 line-height-body-5"
      >
        A notification email will be sent to the requester when you complete this action. If you would like, you may also send a copy to the TRB mailbox and/or to any additional attendees.
      </p>
      <div
        class="usa-form-group margin-top-4"
      >
        <legend>
          Choose recipients
        </legend>
        <p
          class="margin-bottom-0 margin-top-05"
        >
          <span
            class="text-bold"
          >
            2
          </span>
           recipients selected
        </p>
        <ul
          class="usa-list usa-list--unstyled"
        >
          <li>
            <div
              class="easi-checkbox usa-checkbox"
            >
              <input
                checked=""
                class="usa-checkbox__input"
                id="notifyEuaIds.0"
                name="notifyEuaIds"
                type="checkbox"
                value="ABCD"
              />
              <label
                class="usa-checkbox__label"
                for="notifyEuaIds.0"
              >
                Adeline Aarons, CMS (Requester)
              </label>
            </div>
          </li>
          <li>
            <div
              class="easi-checkbox usa-checkbox"
            >
              <input
                checked=""
                class="usa-checkbox__input"
                id="copyTrbMailbox"
                name="copyTrbMailbox"
                type="checkbox"
                value="true"
              />
              <label
                class="usa-checkbox__label"
                for="copyTrbMailbox"
              >
                Copy TRB Mailbox
              </label>
            </div>
          </li>
          <button
            class="usa-button usa-button--unstyled truncated-content-button display-flex flex-align-center margin-top-2"
            data-testid="truncatedContentButton"
            type="button"
          >
            <svg
              class="usa-icon margin-x-05"
              focusable="false"
              height="1em"
              role="img"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="m7 10 5 5 5-5z"
              />
            </svg>
             Show 3 more recipients
          </button>
        </ul>
      </div>
      <div
        class="margin-top-5"
      >
        <ul
          class="usa-button-group"
        >
          <li
            class="usa-button-group__item"
          >
            <div
              class="display-flex flex-align-center"
            >
              <button
                class="usa-button margin-top-0"
                data-testid="button"
                disabled=""
                type="submit"
              >
                Complete action
              </button>
            </div>
          </li>
        </ul>
        <button
          class="usa-button usa-button--unstyled margin-top-2 display-flex flex-align-center"
          data-testid="button"
          type="button"
        >
          <svg
            class="usa-icon margin-right-05"
            focusable="false"
            height="1em"
            role="img"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
            />
          </svg>
          Cancel action and return to request
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
